packages/arb-validator-core/gobridge/goEth.go

```diff
@@ -0,0 +1,370 @@
+	state                   EthState
+	gracePeriod             common.TimeTicks
+	maxSteps                uint64
+	maxTimeBoundsWidth      uint64
+	arbGasSpeedLimitPerTick uint64
+	escrowRequired          *big.Int
+	owner                   common.Address
+	events                  map[*common.BlockId][]arbbridge.Event
+	creation                *common.BlockId
+	stakers                 map[common.Address]*staker
+	leaves                  map[common.Hash]bool
+	lastConfirmed           common.Hash
+	contractAddress         common.Address
+	nextConfirmed           common.Hash
+}
+
+type challengeData struct {
+	sync.Mutex
+	deadline             common.TimeTicks
+	challengerDataHash   common.Hash
+	state                int
+	challengePeriodTicks common.TimeTicks
+	asserter             common.Address
+	challenger           common.Address
+	challengeType        *big.Int
+}
+
+type pendingTopChallengeData struct {
+	challengeData
+}
+
+type messagesChallengeData struct {
+	challengeData
+}
+
+type execChallengeData struct {
+	challengeData
+}
+
+type staker struct {
+	location           common.Hash
+	creationTimeBlocks *common.TimeBlocks
+	inChallenge        bool
+	balance            *big.Int
+}
+
+type nodeGraph struct {
+	stakeRequirement *big.Int
+	stakers          map[common.Address]*staker
+	leaves           map[common.Hash]bool
+	lastConfirmed    common.Hash
+}
+
+type goMaybeEvent struct {
+	challenge *challengeData
+	event     arbbridge.MaybeEvent
+}
+
+type inbox struct {
+	value common.Hash
+	count *big.Int
+}
+
+func (ib *inbox) addMessageToInbox(msg common.Hash) {
+	ib.value = hashing.SoliditySHA3(
+		hashing.Bytes32(ib.value),
+		hashing.Bytes32(msg),
+	)
+	ib.count = new(big.Int).Add(ib.count, big.NewInt(1))
+}
+
+type void struct{}
+
+var Void void
+
+// goEthData one per 'URL'
+type goEthdata struct {
+	blockMutex             sync.Mutex
+	msgMutex               sync.Mutex
+	Vm                     map[common.Address]*VmData
+	channels               map[common.Address]*channelData
+	rollups                map[common.Address]*rollupData // contract address to rollup
+	challenges             map[common.Address]*challengeData
+	balances               map[common.Address]*big.Int
+	challengeWatchersMutex sync.Mutex
+	challengeWatcherEvents map[*challengeData]map[*common.BlockId][]arbbridge.Event
+	arbFactory             common.Address // eth address to factory address
+	nextAddress            common.Address // unique 'address'
+	nextMsgs               map[*common.BlockId][]goMaybeEvent
+	BlockNumber            uint64
+	inbox                  map[common.Address]*inbox
+	NextBlock              *common.BlockId
+	LastMinedBlock         *common.BlockId
+	//LatestHeight *big.Int
+	blockNumbers map[uint64]*common.BlockId      // block height to blockId
+	blockHashes  map[common.Hash]*common.BlockId // block hash to blockId
+	parentHashes map[common.BlockId]common.Hash  // blokcId to block hash
+	ethWallet    map[common.Address]*big.Int
```

Comment:
What's the difference between `ethWallet` and `balances`?

Reply:
still need to review token and ethWallet functionality